Dynamo: Use out-of-place binary ops instead of in-place

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95446

Fixes issues with things like:
```python
x = 2
x += y.shape[0]
```

resulting in invalid `2 += y.shape[0]` code in the FX graph.

Fix: Whenever dynamic shapes are involved, insert the out-of-place op to the FX graph instead of the in-place op.

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

jbschlosser committed:
Dynamo: Use out-of-place binary ops instead of in-place

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -141,47 +141,32 @@ def _fx_graph_functions():
+            # operator.rshift: (["__rshift__", "__rrshift__", "__irshift__"], operator.irshift),
+            # operator.and_: (["__and__", "__rand__", "__iand__"], operator.iand),
+            # operator.or_: (["__or__", "__ror__", "__ior__"], operator.ior),
+            # operator.xor: (["__xor__", "__rxor__", "__ixor__"], operator.ixor),
```
delete this when done?

jbschlosser replied:
Deleted the code, but left the list of currently unsupported ops with a note for completeness. I could be convinced to get rid of that too if you don't think it's useful

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -141,47 +141,32 @@ def _fx_graph_functions():
+            ),
+            operator.mod: (["__mod__", "__rmod__", "__imod__"], operator.imod),
+            pow: (["__pow__", "__rpow__", "__ipow__"], operator.ipow),
+            operator.pow: (["__pow__", "__rpow__", "__ipow__"], operator.ipow),
```
fwiw, these are defined very regularly, so you could probably also do this by string formatting and getattr (no action needed)

jbschlosser replied:
True, although if we eventually support logical bitwise ops and divmod, this adds a couple edge cases

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -187,6 +187,24 @@ def fn(x):
+            # Test reversal by putting int first
+            y = 2
+            y += p
+            y -= p
```
It's not clear to me that the latter tests necessarily do anything, because at that point y is no longer a constant and so it won't get printed inline.

jbschlosser replied:
You're right - changed this to ensure we're working with a constant each time.

ezyang reviewed (approved):
some Qs though

jbschlosser committed:
Update on "Dynamo: Use out-of-place binary ops instead of in-place"


Fixes issues with things like:
```python
x = 2
x += y.shape[0]
```

resulting in invalid `2 += y.shape[0]` code in the FX graph.

Fix: Whenever dynamic shapes are involved, insert the out-of-place op to the FX graph instead of the in-place op.

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
